cmd/build/v2/build.go

```diff
@@ -143,25 +175,49 @@ func NewBuilderFromScratch(ioCtrl *io.Controller, onBuildFinish []OnBuildFinish)
 	gitRepo := repository.NewRepository(wd)
 	config := getConfig(reg, gitRepo, ioCtrl.Logger())
 
-	buildEnvs := map[string]string{}
-	buildEnvs[OktetoEnableSmartBuildEnvVar] = strconv.FormatBool(config.isSmartBuildsEnable)
 	okCtx := &okteto.ContextStateless{
 		Store: okteto.GetContextStore(),
 	}
 
-	return &OktetoBuilder{
-		Builder:           basic.Builder{BuildRunner: builder, IoCtrl: ioCtrl},
-		Registry:          reg,
-		buildEnvironments: buildEnvs,
-		Config:            config,
-		ioCtrl:            ioCtrl,
-		smartBuildCtrl:    smartbuild.NewSmartBuildCtrl(gitRepo, reg, config.fs, ioCtrl, wdCtrl),
-		oktetoContext:     okCtx,
+	smartBuildCtrl := smartbuild.NewSmartBuildCtrl(gitRepo, reg, config.fs, ioCtrl, wdCtrl)
+	tagger := newImageTagger(config, smartBuildCtrl)
+	metadataCollector := metadata.NewMetadataCollector(okCtx, config, smartBuildCtrl, ioCtrl)
+	imageCtrl := registry.NewImageCtrl(okCtx)
+
+	serviceEnvVarsSetter := environment.NewServiceEnvVarsSetter(ioCtrl, reg)
+	serviceEnvVarsSetter.SetEnvVar(OktetoEnableSmartBuildEnvVar, strconv.FormatBool(config.isSmartBuildsEnable))
+
+	imageChecker := checker.NewImageCacheChecker(
+		okCtx.GetNamespace(),
+		okCtx.GetRegistryURL(),
+		tagger,
+		smartBuildCtrl,
+		imageCtrl,
+		metadataCollector,
+		reg,
+		ioCtrl,
+		serviceEnvVarsSetter)
 
-		onBuildFinish: onBuildFinish,
+	return &OktetoBuilder{
+		Builder:              basic.Builder{BuildRunner: builder, IoCtrl: ioCtrl},
+		Registry:             reg,
+		imageChecker:         imageChecker,
+		Config:               config,
+		ioCtrl:               ioCtrl,
+		smartBuildCtrl:       smartBuildCtrl,
+		oktetoContext:        okCtx,
+		metadataCollector:    metadataCollector,
+		onBuildFinish:        onBuildFinish,
+		serviceEnvVarsSetter: serviceEnvVarsSetter,
+		tagger:               tagger,
 	}
 }
 
+// GetBuildEnvVars returns the build env vars
+func (ob *OktetoBuilder) GetBuildEnvVars() map[string]string {
+	return ob.serviceEnvVarsSetter.GetBuildEnvVars()
```

Comment:
Nit: It sounds odd to have something that is a setter, but then, you call a get. If it handles both get and set, you can call it hendler or manager